### DISCUSSION TITLE
stuttering sound on esp32 wrover devboard with pcm51xx connected to it.

### DISCUSSION THREAD

**Discussion by lazysalmon** [pschatzmann/arduino-snapclient#6] [Q&A]
First of all, thanks so much for all the work in this project.

I've started out creating a platformio project in vscode, copying both the arduino-snapclient and arduino-audio-tools in the src folder of the project. 

i simply copy the example from the arduino-snapclient and paste it into main.cpp, replacing cfg by config in this bit:

` // setup I2S to define custom pins
  auto config = out.defaultConfig();
  config.pin_bck = 14;
  config.pin_ws = 15;
  config.pin_data = 22;
  out.begin(config);`

when compiling, the compiler complains the vector library is missing in the file src/api/SnapProtocol.h, so i add it: 

`#include <vector>`

in Snapconfig.h i leave everything as is, but replace the wifi setup and server ip with the one in my setup. Also give the client a pretty name.

Everything compiles nicely.

I make sure to set my snapserver codec to pcm and try various bitrates, now set to 44100:16:2

i made the bridges on PCM51xx board: 1L, 2L, 3H, 4L and soldered the SCK BCK bridge and connected like:

PCM51xx&nbsp;&nbsp;&nbsp;&nbsp;ESP32
VIN&nbsp;&nbsp;&nbsp;&nbsp;3.3V
GND&nbsp;&nbsp;&nbsp;&nbsp;GND
LCK&nbsp;&nbsp;&nbsp;&nbsp;15
DIN&nbsp;&nbsp;&nbsp;&nbsp;22
BCK&nbsp;&nbsp;&nbsp;&nbsp;14
SCK&nbsp;&nbsp;&nbsp;&nbsp;GND (or unconnected does not seem to make a difference, due to the bridge i assume)

Everything connects and music comes trough when i play on my server, but the sound stutters.

I've tried changing buffer settings, bitrate (even to 8000:16:2), changing pin layouts, tried with an unsoldered PCM51xx, varying different parameters in Snapconfig.h (which i don't fully understand) and also tried with a wroom module (setting psram to false), but that almost is unfunctional.

I also tried adding either 
`//static SnapTimeSyncDynamic synch(50, 50);
static SnapTimeSyncFixed synch(123, 1.0);`

and adding 

`client.begin(synch);`

but that doesn't make a difference

Also when i import the PCM51xx library from https://github.com/tommag/PCM51xx_Arduino/tree/master

and change the code to:

`  // start snap client
  client.begin(synch);
  if (bla.begin(PCM51xx::SAMPLE_RATE_44_1K, PCM51xx::BITS_PER_SAMPLE_16))
  Serial.println("PCM51xx initialized successfully.");`
  
  i get a lot of: 
  
  `[  4311][E][Wire.cpp:422] beginTransmission(): could not acquire lock
[  4312][E][Wire.cpp:526] write(): NULL TX buffer pointer
[  4312][E][Wire.cpp:526] write(): NULL TX buffer pointer
[  4317][E][Wire.cpp:448] endTransmission(): NULL TX buffer pointer
[  4323][E][Wire.cpp:422] beginTransmission(): could not acquire lock
[  4329][E][Wire.cpp:526] write(): NULL TX buffer pointer
[  4334][E][Wire.cpp:526] write(): NULL TX buffer pointer
[  4339][E][Wire.cpp:448] endTransmission(): NULL TX buffer pointer
[  4345][E][Wire.cpp:422] beginTransmission(): could not acquire lock
[  4351][E][Wire.cpp:526] write(): NULL TX buffer pointer
[  4356][E][Wire.cpp:526] write(): NULL TX buffer pointer
[  4362][E][Wire.cpp:448] endTransmission(): NULL TX buffer pointer`

any directives why the stutters happen? 

When i compile the https://github.com/pschatzmann/ESP32-A2DP

everything works so fast and smooth it's almost scary.

ps: i can't wrap my head around:

`#ifndef CONFIG_PROCESSING_TIME_MS 
#  define CONFIG_PROCESSING_TIME_MS 172
#endif
#ifndef CONFIG_STREAMIN_DECODER_BUFFER
#  define CONFIG_STREAMIN_DECODER_BUFFER (12 * 1024)`

What does the 172 do? Also, why is 1024 multiplied by 12?

and lastly:

`#ifndef CONFIG_SNAPCAST_BUFF_LEN 
#  define CONFIG_SNAPCAST_BUFF_LEN 1024`

should 1024 be 1000 if my snapserver buffer length is 1000?

thanks a lot!

**Comment by pschatzmann** [comment]
Not sure what the issue is. I suggest that you use the opus decoder instead of wav.
You could also try to reduce the sampling rate to reduce the data that needs to be transmitted.
Also make sure that you have a good Wifi connection...

The time synchronization is only relevant when you want to play multiple devices at the same time...

I am not using any wire functionality, so there should not be any conflict from my code. Just make sure that you call begin() and define the pins and frequency when you don't want to use the default settings.

**Reply by lazysalmon** [reply]
thank you for the quick reply.

I tried lowering the sample rate to 8000:16:2 and still faced the same issues. I also messed around a bit with other codecs but there i got some errors. I will try again tomorrow probably. 

wifi should be ok, as my other devices stream without problems. These are raspberry pi zeroes however.

So as i'm using multiple devices at the same time, does it suffice to just activate ntp? Or are there other timing settings to tweak?

And for the external sound card, if i understand correctly it's not neccesary to specify which type it is, just the pinouts?

Thanks,

**Comment by pschatzmann** [comment]
NTP should not be necessary and I expect that it would not make any difference.

If you have an [external DAC from this list](https://github.com/pschatzmann/arduino-audio-tools/wiki/External-DAC),  no configuration is necessary.

Which DAC do you have exactly?

**Comment by pschatzmann** [comment]
I guess your issue might also be coming from [the logging](https://github.com/pschatzmann/arduino-snapclient/issues/7)
I just repeated my tests with the opus codec and the quality is perfect!
